vng/vng_test.go

```diff
@@ -0,0 +1,378 @@
+package vng_test
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"math"
+	"net/netip"
+	"testing"
+
+	"github.com/brimdata/zed"
+	"github.com/brimdata/zed/pkg/nano"
+	"github.com/brimdata/zed/pkg/units"
+	"github.com/brimdata/zed/vng"
+	"github.com/brimdata/zed/zcode"
+	"github.com/brimdata/zed/zio/vngio"
+	"github.com/brimdata/zed/zio/zngio"
+	"github.com/brimdata/zed/zson"
+)
+
+type mockFile struct {
+	bytes.Buffer
+}
+
+func (f *mockFile) Close() error { return nil }
+
+func FuzzVngRoundtripGen(f *testing.F) {
+	f.Fuzz(func(t *testing.T, b []byte) {
+		bytesReader := bytes.NewReader(b)
+		context := zed.NewContext()
+		types := genTypes(bytesReader, context, 3)
+		values := genValues(bytesReader, context, types)
+		ColumnThresh := int(binary.LittleEndian.Uint64(genBytes(bytesReader, 8)))
+		if ColumnThresh == 0 {
+			ColumnThresh = 1
+		}
+		if ColumnThresh > vng.MaxSegmentThresh {
+			ColumnThresh = vng.MaxSegmentThresh
+		}
+		SkewThresh := int(binary.LittleEndian.Uint64(genBytes(bytesReader, 8)))
+		if SkewThresh == 0 {
+			SkewThresh = 1
+		}
+		if SkewThresh > vng.MaxSkewThresh {
+			SkewThresh = vng.MaxSkewThresh
+		}
+		writerOpts := vngio.WriterOpts{
+			ColumnThresh: units.Bytes(ColumnThresh),
+			SkewThresh:   units.Bytes(SkewThresh),
+		}
+		roundtrip(t, values, writerOpts)
+	})
+}
+
+func FuzzVngRoundtripBytes(f *testing.F) {
+	f.Fuzz(func(t *testing.T, b []byte) {
+		bytesReader := bytes.NewReader(b)
+		context := zed.NewContext()
+		reader := zngio.NewReader(context, bytesReader)
+		defer reader.Close()
+		values := make([]zed.Value, 0)
+		for {
+			value, err := reader.Read()
+			if err != nil {
+				return
+			}
+			if value == nil {
+				break
+			}
+			values = append(values, *(value.Copy()))
+		}
+		writerOpts := vngio.WriterOpts{
+			ColumnThresh: units.Bytes(vngio.DefaultColumnThresh),
+			SkewThresh:   units.Bytes(vngio.DefaultSkewThresh),
+		}
+		roundtrip(t, values, writerOpts)
+	})
+}
+
+func roundtrip(t *testing.T, valuesIn []zed.Value, writerOpts vngio.WriterOpts) {
+	// Debug
+	//for i := range valuesIn {
+	//    t.Logf("value: in[%v].Bytes()=%v", i, valuesIn[i].Bytes())
+	//    t.Logf("value: in[%v]=%v", i, zson.String(&valuesIn[i]))
+	//}
+
+	// Write
+	var fileIn mockFile
+	writer, err := vngio.NewWriter(&fileIn, writerOpts)
```

Comment:
Use zio.NopCloser instead of mockFile.
```suggestion
	var buf bytes.Buffer
	writer, err := vngio.NewWriter(zio.NopCloser(&buf), writerOpts)
```